File: spoj/95-STPAR/solution_95.py
def can_arrange_trucks(trucks):
    nextTarget = 1
    i = 0
    stack = []

    while i < len(trucks) and nextTarget <= len(trucks):
        if nextTarget == trucks[i]:
            nextTarget += 1
            i += 1
        elif len(stack) != 0 and stack[-1] == nextTarget:
            stack.pop()
            nextTarget += 1
        elif len(stack) == 0 or stack[-1] >= nextTarget:
            stack.append(trucks[i])
            i += 1
        else:
            return "no"

    for x in stack[::-1]:
        if x != nextTarget:
            return "no"
        nextTarget += 1
    return "yes" if (nextTarget - 1) == len(trucks) else "no"


if __name__ == "__main__":
    while True:
        n = int(input())
        if n == 0:
            break
        trucks = list(map(int, input().strip().split(" ")))
        print(can_arrange_trucks(trucks))

File: spoj/95-STPAR/solution_95.cpp
#include <iostream>
using namespace std;

string canArrangeTrucks(int *trucks, int size)
{
    int top = -1;
    int nextEntry = 1;
    int stack[size];
    int i = 0;
    while (nextEntry <= size && i < size)
    {
        if (trucks[i] == nextEntry)
        {
            nextEntry++;
            i++;
        }
        else
        {
            if (top != -1 && stack[top] == nextEntry)
            {
                nextEntry++;
                top--;
            }
            else if (top != -1 && stack[top] < trucks[i])
                return "no";

            else
            {
                stack[++top] = trucks[i];
                i++;
            }
        }
    }
    for (i = top; i >= 0; i--)
    {
        if (nextEntry != stack[top--])
        {
            return "no";
        }
        nextEntry++;
    }
    return ((nextEntry - 1) != size) ? "no" : "yes";
}

int main()
{
    while (true)
    {
        int n;
        cin >> n;

        if (n == 0)
            break;

        int trucks[n];
        for (int i = 0; i < n; i++)
            cin >> trucks[i];
        cout << canArrangeTrucks(trucks, n) << "\n";
    }

    return 0;
}